In [1]:
import pandas as pd
from sklearn.impute import SimpleImputer
import numpy as np

In [2]:
data = pd.read_csv('./dataset/hotel_bookings.csv')
data

hotel  is_canceled  lead_time  arrival_date_year  \
0       Resort Hotel            0        342               2015   
1       Resort Hotel            0        737               2015   
2       Resort Hotel            0          7               2015   
3       Resort Hotel            0         13               2015   
4       Resort Hotel            0         14               2015   
...              ...          ...        ...                ...   
119385    City Hotel            0         23               2017   
119386    City Hotel            0        102               2017   
119387    City Hotel            0         34               2017   
119388    City Hotel            0        109               2017   
119389    City Hotel            0        205               2017   

       arrival_date_month  arrival_date_week_number  \
0                    July                        27   
1                    July                        27   
2                    July                        27   
3                    July                        27   
4                    July                        27   
...                   ...                       ...   
119385             August                        35   
119386             August                        35   
119387             August                        35   
119388             August                        35   
119389             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
0                               1                        0   
1                               1                        0   
2                               1                        0   
3                               1                        0   
4                               1                        0   
...                           ...                      ...   
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        stays_in_week_nights  adults  ...  deposit_type  agent company  \
0                          0       2  ...    No Deposit    NaN     NaN   
1                          0       2  ...    No Deposit    NaN     NaN   
2                          1       1  ...    No Deposit    NaN     NaN   
3                          1       1  ...    No Deposit  304.0     NaN   
4                          2       2  ...    No Deposit  240.0     NaN   
...                      ...     ...  ...           ...    ...     ...   
119385                     5       2  ...    No Deposit  394.0     NaN   
119386                     5       3  ...    No Deposit    9.0     NaN   
119387                     5       2  ...    No Deposit    9.0     NaN   
119388                     5       2  ...    No Deposit   89.0     NaN   
119389                     7       2  ...    No Deposit    9.0     NaN   

       days_in_waiting_list customer_type     adr  \
0                         0     Transient    0.00   
1                         0     Transient    0.00   
2                         0     Transient   75.00   
3                         0     Transient   75.00   
4                         0     Transient   98.00   
...                     ...           ...     ...   
119385                    0     Transient   96.14   
119386                    0     Transient  225.43   
119387                    0     Transient  157.71   
119388                    0     Transient  104.40   
119389                    0     Transient  151.20   

        required_car_parking_spaces  total_of_special_requests  \
0                                 0                          0   
1                                 0                          0   
2                                 0                          0   
3             

In [3]:
data.columns

Index(['hotel', 'is_canceled', 'lead_time', 'arrival_date_year',
       'arrival_date_month', 'arrival_date_week_number',
       'arrival_date_day_of_month', 'stays_in_weekend_nights',
       'stays_in_week_nights', 'adults', 'children', 'babies', 'meal',
       'country', 'market_segment', 'distribution_channel',
       'is_repeated_guest', 'previous_cancellations',
       'previous_bookings_not_canceled', 'reserved_room_type',
       'assigned_room_type', 'booking_changes', 'deposit_type', 'agent',
       'company', 'days_in_waiting_list', 'customer_type', 'adr',
       'required_car_parking_spaces', 'total_of_special_requests',
       'reservation_status', 'reservation_status_date'],
      dtype='object')

In [ ]:
"""
칼럼 삭제 목록 
agent
country
company
reservation_status
reservation_status_date
distribution_channel: 예약채널
market_segment : 시장세분화
"""

In [4]:
data = data.drop(columns=['agent','country','company','reservation_status_date','reservation_status','distribution_channel','market_segment'])
data.columns

Index(['hotel', 'is_canceled', 'lead_time', 'arrival_date_year',
       'arrival_date_month', 'arrival_date_week_number',
       'arrival_date_day_of_month', 'stays_in_weekend_nights',
       'stays_in_week_nights', 'adults', 'children', 'babies', 'meal',
       'is_repeated_guest', 'previous_cancellations',
       'previous_bookings_not_canceled', 'reserved_room_type',
       'assigned_room_type', 'booking_changes', 'deposit_type',
       'days_in_waiting_list', 'customer_type', 'adr',
       'required_car_parking_spaces', 'total_of_special_requests'],
      dtype='object')

In [21]:
canceled = data[data['is_canceled']== 1]
not_canceled = data[data['is_canceled']==0]

In [24]:
can = canceled.select_dtypes(exclude='object')
notcan = not_canceled.select_dtypes(exclude='object')

In [60]:
can.columns

Index(['is_canceled', 'lead_time', 'arrival_date_year',
       'arrival_date_week_number', 'arrival_date_day_of_month',
       'stays_in_weekend_nights', 'stays_in_week_nights', 'adults', 'children',
       'babies', 'is_repeated_guest', 'previous_cancellations',
       'previous_bookings_not_canceled', 'booking_changes',
       'days_in_waiting_list', 'adr', 'required_car_parking_spaces',
       'total_of_special_requests'],
      dtype='object')

In [61]:
notcan.columns

Index(['is_canceled', 'lead_time', 'arrival_date_year',
       'arrival_date_week_number', 'arrival_date_day_of_month',
       'stays_in_weekend_nights', 'stays_in_week_nights', 'adults', 'children',
       'babies', 'is_repeated_guest', 'previous_cancellations',
       'previous_bookings_not_canceled', 'booking_changes',
       'days_in_waiting_list', 'adr', 'required_car_parking_spaces',
       'total_of_special_requests'],
      dtype='object')

In [34]:
#정규성 검사 
from scipy.stats import normaltest
normaltest(can.lead_time)

NormaltestResult(statistic=6077.828068850519, pvalue=0.0)

In [38]:
# t검정 시행- one sample t-test 시행
from scipy import stats
stats.ttest_ind(can['lead_time'], notcan['lead_time'],equal_var = False)

Ttest_indResult(statistic=99.07475331704227, pvalue=0.0)

In [25]:
normaltest(notcan)

NormaltestResult(statistic=array([186556.38227241,  20031.27882389,  15639.82285424,  20889.08157475,
        59140.19868617,  26058.25559403,  51132.64894647,   2964.55905092,
        61027.71179189, 177782.43389112,  65822.8263883 , 192088.56716066,
       162752.73327057,  82687.00348762, 131224.84016262,  11944.21159321,
        52744.00600143,  11184.18905303]), pvalue=array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0.]))

<strong> - deposit_type : 범주형이라 t검정 시행 불가

---

독립표본 (independent) t-test: 두 집단이 서로 다른 모집단에서 추출된, 독립된 두 집단의 평균의 차이를 검정하는 것
    => 독립표본 t 검정 시행

<strong>귀무가설 :두 집단의 모평균(모집단의 평균)은 동일하다 <br>
<strong>대립가설: 두 집단의 모평균(모집단의 평균)은 동일하지 않다.

### lead_time

In [70]:
from scipy import stats
# 등분산 검정
print(stats.levene(can['lead_time'], notcan['lead_time']))
print(stats.fligner(can['lead_time'], notcan['lead_time']))
print(stats.bartlett(can['lead_time'], notcan['lead_time']))

LeveneResult(statistic=3601.8002794448134, pvalue=0.0)
FlignerResult(statistic=5293.5110147009555, pvalue=0.0)
BartlettResult(statistic=4014.982986160867, pvalue=0.0)


<strong> Levene의 등분산 검정 결과 F값이 3601.800로 확인됐다. F값이란 집단간 분산/집단내 분산을 말한다. 따라서 F값이 크면클수록
두 표본의 분산이 동일하지 않다는 뜻이다.

In [44]:
#등분산이 아닌 독립표본 t-검정 실행 
result = stats.ttest_ind(can.lead_time, notcan.lead_time, equal_var=False) 
print('t statistic : %.3f \np-value : %.3f' % (result))

t statistic : 99.075 
p-value : 0.000


 <strong>독립표본 t-검정 결과 t값이 99.075, p-value가 0.000으로 신뢰수준 99%하에서 "lead_time에 따른 예약취소여부엔 차이가 없다."라는 가설을 기각한다.
따라서 Retailer type에 따른 평균 매출액엔 차이가 있는 것으로 확인됐다

### booking_changes

In [46]:
from scipy import stats
# 등분산 검정
print(stats.levene(can['booking_changes'], notcan['booking_changes']))
print(stats.fligner(can['booking_changes'], notcan['booking_changes']))
print(stats.bartlett(can['booking_changes'], notcan['booking_changes']))

LeveneResult(statistic=2541.731400613089, pvalue=0.0)
FlignerResult(statistic=3650.908800115861, pvalue=0.0)
BartlettResult(statistic=11923.175380714158, pvalue=0.0)


In [51]:
#등분산이 아닌 독립표본 t-검정 실행 
result = stats.ttest_ind(can.booking_changes, notcan.booking_changes, equal_var=False) 
print('t statistic : %.3f \np-value : %.3f' % (result))

t statistic : -56.746 
p-value : 0.000


<strong>결과 해석 : 예약을 취소한 집단과 취소하지 않은 집단간의 분산의 차이가 없다, 즉 동일하다고 할수 있다.

### required_car_parking_spaces

In [47]:
from scipy import stats
# 등분산 검정
print(stats.levene(can['required_car_parking_spaces'], notcan['required_car_parking_spaces']))
print(stats.fligner(can['required_car_parking_spaces'], notcan['required_car_parking_spaces']))
print(stats.bartlett(can['required_car_parking_spaces'], notcan['required_car_parking_spaces']))

LeveneResult(statistic=4744.260084629089, pvalue=0.0)
FlignerResult(statistic=4625.534246549402, pvalue=0.0)
BartlettResult(statistic=inf, pvalue=0.0)


C:\Users\medici\anaconda3\lib\site-packages\scipy\stats\morestats.py:2467: RuntimeWarning: divide by zero encountered in log
  numer = (Ntot*1.0 - k) * log(spsq) - np.sum((Ni - 1.0)*log(ssq), axis=0)


In [52]:
#등분산이 아닌 독립표본 t-검정 실행 
result = stats.ttest_ind(can.required_car_parking_spaces, notcan.required_car_parking_spaces, equal_var=False) 
print('t statistic : %.3f \np-value : %.3f' % (result))

t statistic : -89.798 
p-value : 0.000


### previous_cancellations

In [49]:
from scipy import stats
# 등분산 검정
print(stats.levene(can['previous_cancellations'], notcan['previous_cancellations']))
print(stats.fligner(can['previous_cancellations'], notcan['previous_cancellations']))
print(stats.bartlett(can['previous_cancellations'], notcan['previous_cancellations']))

LeveneResult(statistic=1465.864983065514, pvalue=8.93266e-319)
FlignerResult(statistic=8117.859348505634, pvalue=0.0)
BartlettResult(statistic=128252.0028606421, pvalue=0.0)


In [53]:
#등분산이 아닌 독립표본 t-검정 실행 
result = stats.ttest_ind(can.previous_cancellations, notcan.previous_cancellations, equal_var=False) 
print('t statistic : %.3f \np-value : %.3f' % (result))

t statistic : 30.026 
p-value : 0.000


### previous_bookings_not_canceled

In [50]:
from scipy import stats
# 등분산 검정
print(stats.levene(can['previous_bookings_not_canceled'], notcan['previous_bookings_not_canceled']))
print(stats.fligner(can['previous_bookings_not_canceled'], notcan['previous_bookings_not_canceled']))
print(stats.bartlett(can['previous_bookings_not_canceled'], notcan['previous_bookings_not_canceled']))

LeveneResult(statistic=394.07204580517646, pvalue=1.4894651920729486e-87)
FlignerResult(statistic=1530.0476412047547, pvalue=0.0)
BartlettResult(statistic=41008.23889721833, pvalue=0.0)


In [54]:
#등분산이 아닌 독립표본 t-검정 실행 
result = stats.ttest_ind(can.previous_bookings_not_canceled, notcan.previous_bookings_not_canceled, equal_var=False) 
print('t statistic : %.3f \np-value : %.3f' % (result))

t statistic : -24.194 
p-value : 0.000


<strong> t검정 수행 못하는 항목 :
    customer_type: 범주형
    ,assigned_room_type : 범주형
    

<strong>결론 : 5개의 독립변수 모두 p-value 값이 0 이기때문에 p값이 0.05보다 작기때문에 귀무가설 기각한다.